### PR TITLE
Add fallback vnet for failover dns resolving.

### PR DESCRIPTION
**Pull request by Acenl12**
# Overview/Summary
Add fallback vnet for failover dns resolving.

### CODE REVIEW

**Comment by Acenl12** [comment]
@microsoft-github-policy-service agree

File: infra-as-code/bicep/modules/privateDnsZones/privateDnsZones.bicep [hunk 1]
```diff
@@ -180,6 +184,18 @@ resource resVirtualNetworkLink 'Microsoft.Network/privateDnsZones/virtualNetwork
   dependsOn: resPrivateDnsZones
 }]
 
+resource fallbackresbVirtualNetworkLink 'Microsoft.Network/privateDnsZones/virtualNetworkLinks@2020-06-01' = [for privateDnsZoneName in varPrivateDnsZonesMerge: if (!empty(parVirtualNetworkIdToLink)) {
```
Comment (jtracey93) [inline infra-as-code/bicep/modules/privateDnsZones/privateDnsZones.bicep:187]:
```suggestion
resource resVirtualNetworkLinkFailover 'Microsoft.Network/privateDnsZones/virtualNetworkLinks@2020-06-01' = [for privateDnsZoneName in varPrivateDnsZonesMerge: if (!empty(parVirtualNetworkIdToLinkFailover)) {
```

File: infra-as-code/bicep/modules/privateDnsZones/parameters/privateDnsZones.parameters.all.json [hunk 2]
```diff
@@ -86,6 +86,9 @@
     },
     "parVirtualNetworkIdToLink": {
       "value": "/subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/resourceGroups/xxxxxxxxxxx/providers/Microsoft.Network/virtualNetworks/xxxxxxxxxxx"
+    },
+        "parVirtualNetworkIdToLinkFailover": {
+      "value": "/subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/resourceGroups/xxxxxxxxxxx/providers/Microsoft.Network/virtualNetworks/xxxxxxxxxxx"
```
Comment (jtracey93) [inline infra-as-code/bicep/modules/privateDnsZones/parameters/privateDnsZones.parameters.all.json:91]:
```suggestion
      "value": ""
```

default to empty as not a common architecture for most

File: infra-as-code/bicep/modules/privateDnsZones/privateDnsZones.bicep [hunk 3]
```diff
@@ -83,6 +83,10 @@ param parTags object = {}
 @sys.description('Resource ID of VNet for Private DNS Zone VNet Links.')
 param parVirtualNetworkIdToLink string = ''
 
+@sys.description('Resource ID of VNet for Failover Private DNS Zone VNet Links.')
+param parVirtualNetworkIdToLinkFailover string = ''
```
Comment (jtracey93) [inline infra-as-code/bicep/modules/privateDnsZones/privateDnsZones.bicep:87]: do we consider renaming `Failover` to `Alternate`, as its not really a failover as such?

Would require updates to names in other places :) But think its the right thing to do

**Review by jtracey93 — CHANGES_REQUESTED** [review]
Thanks @Acenl12 for the PR.

A few comments for review please

File: infra-as-code/bicep/modules/privateDnsZones/privateDnsZones.bicep (hunk 3 above)
Comment (Acenl12) [reply infra-as-code/bicep/modules/privateDnsZones/privateDnsZones.bicep:87]: In the documentation of azure private dns resolver for example a second link is documented as failover.

Comment (jtracey93) [reply infra-as-code/bicep/modules/privateDnsZones/privateDnsZones.bicep:87]: Can you share a link?

Comment (Acenl12) [reply infra-as-code/bicep/modules/privateDnsZones/privateDnsZones.bicep:87]: https://learn.microsoft.com/en-us/azure/dns/private-resolver-reliability  @jtracey93

**Review by jtracey93 — CHANGES_REQUESTED** [review]
Thanks for the work here @Acenl12.

Could you please review the comments and suggestion I have left in this review and also the below points:

1. Should we expose this new parameter in the `hubNetworking.bicep` and `vwanConnectivity.bicep` modules? I think we should, if so please add this parameter there also and update docs, parameter files etc.
2. Can you run the script detailed here: https://github.com/Azure/ALZ-Bicep/wiki/Contributing#manually-generating-the-parameter-markdown-files to update the parameter docs as the workflow is failing.

Thanks

**Comment by Acenl12** [comment]
> Thanks for the work here @Acenl12.
> 
> Could you please review the comments and suggestion I have left in this review and also the below points:
> 
> 1. Should we expose this new parameter in the `hubNetworking.bicep` and `vwanConnectivity.bicep` modules? I think we should, if so please add this parameter there also and update docs, parameter files etc.
> 2. Can you run the script detailed here: https://github.com/Azure/ALZ-Bicep/wiki/Contributing#manually-generating-the-parameter-markdown-files to update the parameter docs as the workflow is failing.
> 
> Thanks

I have add the markdown in my branch, hopefully that will be automatically updates in the pull request

**Comment by Acenl12** [comment]
> > Thanks for the work here @Acenl12.
> > Could you please review the comments and suggestion I have left in this review and also the below points:
> > 
> > 1. Should we expose this new parameter in the `hubNetworking.bicep` and `vwanConnectivity.bicep` modules? I think we should, if so please add this parameter there also and update docs, parameter files etc.
> > 2. Can you run the script detailed here: https://github.com/Azure/ALZ-Bicep/wiki/Contributing#manually-generating-the-parameter-markdown-files to update the parameter docs as the workflow is failing.
> > 
> > Thanks
> 
> I have add the markdown in my branch, hopefully that will be automatically updates in the pull request

Why do you think it's a good idea that it also there? I can't judge. I think modules should kept solo with as less references as possible, if somebody only wants to run a certain module. I generated the markdown hopefully that's sufficient now.

**Comment by jtracey93** [comment]
> > > Thanks for the work here @Acenl12.
> > > Could you please review the comments and suggestion I have left in this review and also the below points:
> > > 
> > > 1. Should we expose this new parameter in the `hubNetworking.bicep` and `vwanConnectivity.bicep` modules? I think we should, if so please add this parameter there also and update docs, parameter files etc.
> > > 2. Can you run the script detailed here: [Azure/ALZ-Bicep/wiki/Contributing#manually-generating-the-parameter-markdown-files](https://github.com/Azure/ALZ-Bicep/wiki/Contributing#manually-generating-the-parameter-markdown-files) to update the parameter docs as the workflow is failing.
> > > 
> > > Thanks
> > 
> > 
> > I have add the markdown in my branch, hopefully that will be automatically updates in the pull request
> 
> Why do you think it's a good idea that it also there? I can't judge. I think modules should kept solo with as less references as possible, if somebody only wants to run a certain module. I generated the markdown hopefully that's sufficient now.@jtracey93

Thanks @Acenl12, as for point 1, the primary consumption method for the Private DNS Zones module is the Hub Networking and VWAN Bicep modules as per the deployment flow: https://github.com/Azure/ALZ-Bicep/wiki/DeploymentFlow

So if we can add the parameters to these modules and pass them into the Private DNS Zones module you have updated here, we allow anyone, no matter how they are consuming/using the modules to get access to the new feature you've added in this PR.

Please make the change requested and expose in these modules 👍

Thanks

**Comment by Acenl12** [comment]
@jtracey93  In the hub networking module I don't see any parameter for a private dns zone virtual network link, are you sure the failover needs to be add there? https://github.com/Acenl12/ALZ-Bicep/edit/Acenl12-patch-1/infra-as-code/bicep/modules/hubNetworking/hubNetworking.bicep. I added it to the vwan.

File: infra-as-code/bicep/modules/hubNetworking/hubNetworking.bicep [hunk 4]
```diff
@@ -761,6 +764,7 @@ module modPrivateDnsZones '../privateDnsZones/privateDnsZones.bicep' = if (parPr
     parLocation: parLocation
     parTags: parTags
     parVirtualNetworkIdToLink: resHubVnet.id
+    parVirtualNetworkIdToLinkFailover        : parVirtualNetworkIdToLinkFailover
```
Comment (jtracey93) [inline infra-as-code/bicep/modules/hubNetworking/hubNetworking.bicep:767]:
```suggestion
    parVirtualNetworkIdToLinkFailover: parVirtualNetworkIdToLinkFailover
```

File: infra-as-code/bicep/modules/vwanConnectivity/vwanConnectivity.bicep [hunk 5]
```diff
@@ -337,6 +340,7 @@ module modPrivateDnsZones '../privateDnsZones/privateDnsZones.bicep' = if (parPr
     parPrivateDnsZones: parPrivateDnsZones
     parPrivateDnsZoneAutoMergeAzureBackupZone: parPrivateDnsZoneAutoMergeAzureBackupZone
     parVirtualNetworkIdToLink: parVirtualNetworkIdToLink
+    parVirtualNetworkIdToLinkFailover        : parVirtualNetworkIdToLinkFailover
```
Comment (jtracey93) [inline infra-as-code/bicep/modules/vwanConnectivity/vwanConnectivity.bicep:343]:
```suggestion
    parVirtualNetworkIdToLinkFailover: parVirtualNetworkIdToLinkFailover
```

**Review by jtracey93 — CHANGES_REQUESTED** [review]

File: infra-as-code/bicep/modules/vwanConnectivity/parameters/vwanConnectivity.parameters.all.json [hunk 6]
```diff
@@ -149,6 +149,14 @@
...
       }
     },
+    "parVirtualNetworkIdToLinkFailover": {
+      "value": "/subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/resourceGroups/HUB_Networking_POC/providers/Microsoft.Network/virtualNetworks/alz-hub-eastus-failover"
+    },
+    "parTags": {
+      "value": {
+        "Environment": "Live"
+      }
+    },
```
Comment (jtracey93) [inline infra-as-code/bicep/modules/vwanConnectivity/parameters/vwanConnectivity.parameters.all.json:159]:
```suggestion
```

**Pull request merged**